### PR TITLE
Resolve ENS name in chats and profiles

### PR DESCRIPTION
**Pull request by jeluard**
### Summary

Resolve ENS names defined by user in chats, notifications and profiles.

### Steps to test

- Open Status
- Setup a preferred ENS name
- Verify it's used in local user profile
- Verify it's resolved when sending messages in group chats
- Verify it's resolved when another user opens this user profile

status: ready

### CODE REVIEW

File: src/status_im/transport/message/transit.cljs [hunk 1]
```diff
@@ -132,17 +132,17 @@
...
+                              "c3" (fn [[name profile-image address fcm-token device-info ens-name]]
+                                     (contact/ContactRequestConfirmed. name ens-name profile-image address fcm-token device-info))
                               "c4" (fn [[legacy-content content-type message-type clock-value timestamp content]]
                                      (let [[new-content new-content-type] (legacy->new-message-data (or content legacy-content) content-type)]
                                        (protocol/Message. new-content new-content-type message-type clock-value timestamp)))
                               "c7" (fn [[content content-type message-type clock-value timestamp]]
                                      (protocol/Message. content content-type message-type clock-value timestamp))
                               "c5" (fn [])
-                              "c6" (fn [[name profile-image address fcm-token device-info]]
+                              "c6" (fn [[name ens-name profile-image address fcm-token device-info ens-name]]
```
Comment (cammellos) [inline src/status_im/transport/message/transit.cljs]: seems like it's there twice?

Comment (jeluard) [reply src/status_im/transport/message/transit.cljs]: Good catch!

**Review by vitvly — APPROVED** [review]

**Review by dmitryn — APPROVED** [review]

**Comment by yenda** [comment]
If I understand correctly only added contacts can see your ens name? Is there another PR for public chats?

**Comment by jeluard** [comment]
@yenda In profile yes. In chat messages it's shown if you have the right preference enabled

**Comment by churik** [comment]
@jeluard skip my fix and just rebase, a bit more fix is needed for this test (will do after rebase).

About PR: am I doing something wrong?
1) User1 (IOS):  restore account, that owns ENS username.
- added it in Profile and switched on "Show my ENS username in chats"
- sent several messages in public chat `#m`
2) User2 (Android):
- joined public chat `#m`. Can see default username for User1 - I expect here to see ENS name
- as User1 added  User2 as contact, anyway can see only default username for User1 (if open his profile as well) - I expect here to see ENS name.

Both users are in PR builds on mainnet.
So what exactly should I do to see ENS names in chat?

**Comment by jeluard** [comment]
@churik User1 should:

- register/connect the ENS name
- select it as default
- switch on 'show my ENS username'
- send some messages

Then User2 will see ENS in chats.

If User1 adds some user as contact, the ENS name is shown in profile (`show my ENS userename` is irrelevant here)

**Comment by Serhy** [comment]
@jeluard it is a bit unclear how `Show ENS name in chats` setting conjunct with the users which in 'My contact list' (which chats exactly by the way?:) because at the moment with `Show ENS name in chats` on we share ENS in public, but in 1-1 chat it's still default).

There are many places in app where we showed contact name (for example Login screen, should we display now ENS names there?).

So we created a brief matrix to clarify some scenarios:
https://docs.google.com/spreadsheets/d/1eg-_h5TfPyQIonpMAidz8x-CcVEYW6IAFmj3vSAeRdM/edit#gid=0

Could you please take a look and confirm it's correct? 

Overall summary items from above matrix are:

1) User who exists in my contact list has higher priority than 'Show my ENS in chats'. It's like, if I added someone in contacts I trust this user and no sense to hide my ENS anywhere from this contact.

2) `Show my ENS name setting` impacts only public chat (until I add user in contact list - sin that case item 1 above applies). It's like "I don't mind to share my ENS name for public when I post any message in Public chat, but because I don't want to share my public key for all of the world and get spammed in 1-1 chat - I do not reveal my ENS in my profile and 1-1 chat"

3) Custom userimage (UserPic) regulates only by whether user in my contact list or not (Show ENS name has no effect here)

**Comment by jeluard** [comment]
@Serhy Great work!

Some clarifications:
- `show username` only affects ENS shown in chat (so only for messages)
- `show username` and user in contact are totally independant
- `show username` applies per message (so if you have it on, all messages embed ENS, off they don't)
- this PR doesn't change the image behaviour

> There are many places in app where we showed contact name (for example Login screen, should we display now ENS names there?).

Not sure. @errorists maybe you want to shim in?

Also what is `contact name` for you?

**Comment by errorists** [comment]
Sure. @Serhy Ok so by contact name I guess you're referring to what we describe as `Chosen Name`, the one you get to customise and pick for yourself. [Here's the full diagram by the way](https://discuss.status.im/t/profile-information-structure-please-review/1080/15) 

ENS effectively replaces it. Anywhere you'd previously see your `Chosen Name` you'll get to see 
-> your `@ENS` name, if you have multiple, we display the one that's chosen by you as the `primary` one.
-> if you don't have one, it's the `three word anonymous funny animal totem` :)

**Comment by Serhy** [comment]
Cool, thanks @jeluard, thanks @errorists ! 
The https://discuss.status.im/t/profile-information-structure-please-review/1080/15 explains what we are going to achieve at this stage. 
Own profile ENS name and places where we need to update it will be addressed separately from here.

**Comment by Serhy** [comment]
## Issue 1 
> * Verify it's resolved when another user opens this user profile

ENS name is not resolved for non-contacts when they open this user profile (it's resolved, though for those who in my contact list)


**To reproduce:**
1) Open status and login to account with ENS name already existing (UserA)
2) UserA enables `Show my ENS username in chats` setting
3) UserA joins public chat and sends a message
4) UserB joins the same public chat, receives message from UserA
5) UserB taps UserA's profile icon to navigat to UserA's profile
**Actual result:** Default three-name is displayed to UserB
**Expected result**: ENS name displayed
The same with username values on 1-1 chat screen, could we fix them as well, please?

Screenshot 2019-06-27 at 18 03 42

## Issue 2
Having only one ENS username and enabled `Show my ENS username in chats` has no effect in applying ENS name be shown in chats. 
If there are two or more ENS names - then there is an ability to select `Primary username` between them, - and in that case (case of selecting Primary username) it works. But doesn't work until `Primary username` selected

**To reproduce:**
**Preconditions**: Fresh Status build. 
1) Recover an account which has ENS names already connected to chat key.
2) Navigate to ENS usernames in `Profile` -> `Add username` -> add your ENS username
3) Join any public chat and send text message
4) With different user join the same public chat and verify there is ENS name present in message sent in the 3rd item

**Comment by jeluard** [comment]
Thanks @Serhy !

One is expected and is due to how things are working. In this case the ENS name is tied to a message, not a profile.

I will fix Two.

**Comment by jeluard** [comment]
@Serhy I pushed a fix for 2.

**Comment by Serhy** [comment]
Push notifications do not appear for new/recovered users in the build with current changes. 
However, PNs are working (with old 'Usernames'/default three-name) after upgrade from release on to this PR build (and showing old 'Username', even ENS already applied)

So we need that ENS names would be applied and shown in Push Notifications

**Preconditions**: 
User does not exist yet in fresh PR Status build 
**Steps to reproduce**:
1) Open Status
2) Create new account (UserA) and with another device create another account (UserB)
3) UserA and UserB add both in Contacts and exchange with the messages
4) UserA put Status on background
5) UserB sends a text message to UserA
**Actual result:** No Push Notification appears at all.
**Expected result**: Push Notification appears with UserB default (or ENS if it is set!) name.

Logcat from Android (no obvious errors there though): 
[Status.log](https://github.com/status-im/status-react/files/3339532/Status.log)

**Comment by jeluard** [comment]
> Push notifications do not appear for new/recovered users in the build with current changes.

You mean PN are not sent at all?

**Comment by jeluard** [comment]
@Serhy Can you reproduce the same with nightly? I didn't change anything related to PN in this PR AFAIK

**Comment by Serhy** [comment]
@jeluard No it's was all fine with nightly and can't reproduce with todays (1st of July) nightly build. 
With PR builds I don't 
Let me create a nightly build based on this PR branch and check there (maybe it's PR-build settings specific).

**Comment by Serhy** [comment]
@jeluard the same behaviour on iOS and Android with nightly build created from `ens-resolution` branch (job= https://ci.status.im/job/status-react/job/combined/job/mobile-android/11392/console -> Android http://status-im.ams3.digitaloceanspaces.com/StatusIm-190701-084916-39686a-nightly.apk) - no Push Notifications appears at all when user receives a message from contact which is IN his(her) contact list.
(But the same works fine with todays nightly build)

So as described in https://github.com/status-im/status-react/pull/8487#issuecomment-506745646 -  - reproduced only with the changes made in current PR.

One more important note discovered: 
It's all fine if the account has ENS name setted up (not only just purchased, but setted up as well). 
So the issue for new accounts (or recovered accounts) is that: no ENS username added in account -> PNs won't work for you.

**Comment by jeluard** [comment]
Hmm weird. I rebase just in case, would be nice if you could give it a quick try.

Is it covered by e2e tests?

**Comment by Serhy** [comment]
> Is it covered by e2e tests?

No, we don't have e2e tests for Push Notifications yet. In ToDo list.

**Comment by Serhy** [comment]
> Hmm weird. I rebase just in case, would be nice if you could give it a quick try.

Yes, ok, waiting for build...

**Comment by Serhy** [comment]
@jeluard Checked with https://i.diawi.com/e7HoAP and https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190701-110317-6bf206-pr.apk 

Same issue. No Push Notification appears on any of iOS / Android devices between contacts who in the contact list of each other.

**Comment by Serhy** [comment]
Latest build with more logs tested. Below logs for the scenario (excluding Preconditions part):
Preconditions:
1) Fresh Status build installed
2) UserA has been recovered (no ENS name set up)
3) UserB sent to UserA text message
4) UserA hit 'Add to contacts' to add in contact list UserB

Steps to reproduce:
1) Open Status and log in to UserA account
2) Put Status on background
3) UserB sends several text messages to UserA
**Actual result:** No Push Notification appears

**UserA's logs:**

FullDeviceLogcat: [logcatMORElogs.log](https://github.com/status-im/status-react/files/3349913/logcatMORElogs.log)
SameLogcatFilteredByStatusPID: [logcatMOREfilteredByStatusPID.log](https://github.com/status-im/status-react/files/3349914/logcatMOREfilteredByStatusPID.log)
Geth.log with DEBUG log level: [gethMORElogs.log](https://github.com/status-im/status-react/files/3349915/gethMORElogs.log)

**Comment by Serhy** [comment]
@jeluard works well now. 
PN display ENS (hash when app is closed)
ENS name displays in chats (Public/Group) when `Show my ENS username in chats` setting is ON (and default username is set when this setting is OFF)

**Pull request merged**